Add RFC for JWS-signing PGXN releases

theory opened the pull request.

theory committed:
Add RFC for JWS-signing PGXN releases

pgguru commented on text/0000-release-meta-spec-v2.md:
```diff
@@ -0,0 +1,305 @@
+    is uploaded in some other archive format than zip).
+
+4.  PGXN Manager copies the `META.json` file, constructs the payload for
+    sighing, signs with its public key, then adds the `release` object to the
```
```suggestion
    signing, signs with its public key, then adds the `release` object to the
```

theory replied:
Fixed in fa66f7a, thanks!

pgguru commented on text/0000-release-meta-spec-v2.md:
```diff
@@ -0,0 +1,305 @@
+
+*   This pattern could make it more difficult for clients to install code from
+    PGXN, especially if they incorrectly validate the signature.
+*   Some clients may choose not to implement the validation, potentially
```
Should there be an option to install only trusted packages, or that's a client-specific consideration?

theory replied:
Client-specific. My intention, once we have most stuff signed, would be for the official CLI to validate by default. But we have no day in how others want to use it.

pgguru commented on text/0000-release-meta-spec-v2.md:
```diff
@@ -0,0 +1,305 @@
+    PGXN `META.json` format to enable best-in-class digital signatures and
+    validation.
+*   Without key signing, questions will gradually mount as to the
+    trustworthiness of releases published on PGXN. In the era of [what's teh
```
```suggestion
    trustworthiness of releases published on PGXN. In the era of [what's the
```

(maybe this is intentional?)

theory replied:
Fixed in fa66f7a, thanks!

pgguru commented on text/0000-release-meta-spec-v2.md:
```diff
@@ -0,0 +1,305 @@
+
+## Unresolved questions
+
+None currently.
```
If public keys are distributed through the same distribution method as the packages (i.e., PGXN web services) what prevents the whole system from being compromised?  Seems like a different distribution method (at least from a different host) would alleviate at least some of this concern.

pgguru replied:
Also in order to augment the `META.json` file from the distribution's copy, we'd by definition be changing the hash of the `META.json` file iteself—does there need to be some sort of validation that the only changes to that file have come from the added (JSON) keys themselves?

theory replied:
This RFC doesn't address key management; I'll add that to the unresolved questions.

The file that includes the signature is not included in the distribution file; it's separate. The `META.json` included in the zip file is unchanged from how authors provide it (IIRC), and is defined by #3.

This RFC _extends_ on #3 by adding this additional metadata, and the resulting "release metadata" file is distributed along side the zip file, not in it. This pattern is true for the original PGXN design, too, where each release results in two files:

* `dist/$name/$version/META.json` — the release metadata file
* `dist/$name/$version/$name-$version.zip` — includes the distribution metadata file

I tried to describe that process in the `Publishing a Release` section.

theory replied:
In fa66f7a, I added a bunch of unresolved questions, including about key management. I also went into a bit more detail on these files and how they're not included in the source distribution file.

theory committed:
Make an edit pass, define object structures

Also add some items to the "Unresolved questions" section and a "Future
possibilities" item for author signing.

sjmiller609 commented on text/0000-release-meta-spec-v2.md:
```diff
@@ -0,0 +1,464 @@
+    `dist/pair/0.1.7/META.json`.
+3.  Fetch the release `META.json` file, read in the `releases/pgxn` object,
+    and use PGXN's public key to verify that it was signed by PGXN. Abort with
+    an error if validation fails.
```
We could mention something about how key invalidation is handled, e.g. if PGXN wants to mark a previously published public key as no longer valid. e.g. could just say "PGXN's current public keys" or similar perhaps

theory replied:
Done in 8c8bab2.

sjmiller609 commented on text/0000-release-meta-spec-v2.md:
```diff
@@ -0,0 +1,464 @@
+1.  A root key pair is maintained by PGXN, with the private key kept offline.
+2.  A release key pair is generated and signed by the private root key, with
+    the private key kept in an online vault accessible only to PGXN Manager.
+3.  The public keys for both keys are published by PGXN. Clients should embed
+    the root public key in their sources.
```
It's normal to grab the public key from the trusted entity instead of being embedded, which helps support the case of invalidating keys when necessary, for example invalidating public keys associated to leaked private keys. For example, https://clerk.com/docs/reference/backend-api/tag/JWKS

sjmiller609 replied:
Some clients literally download the public key on every authentication but in my opinion it's sensible to cache it for a duration equal to your key invalidation timeline tolerance.

theory replied:
I might recommend embedding the root public key and downloading the intermediate key used for signing.

theory replied:
Oh, that line already mentions the root key. I expanded it in 8c8bab2 to read:

> 3.  The public keys for both keys are published by PGXN. Clients should embed the root public key in their sources but regularly fetch and validate signing keys against it.

sjmiller609 commented on text/0000-release-meta-spec-v2.md:
```diff
@@ -0,0 +1,464 @@
+wheel], and is enabled by the separation of the release `META.json` file from
+the release file it describes.
+
+The use of [JWS] ensures a widely-vetted key signing and distribution
+standard, and the likelihood that clients can take advantage of well-tested,
+mature libraries to handle signing and validation. And finally, its design
+allows for key rotation when necessary.
```
I might have missed something about key rotation, but I'm not understanding how that works

theory replied:
I think it's not JWS that allows it so much as the support for multiple signatures enabled by JWS JSON Serialization: one could sign with both the old a new key for a bit, for example.

theory replied:
Updated that last sentence in 8c8bab2 to:

> And finally, the use of RFC 7515-standard [JWS JSON Serialization] allows multiple signatures, which will simplify key rotation.

sjmiller609 commented on text/0000-release-meta-spec-v2.md:
```diff
@@ -0,0 +1,464 @@
+       "payload": "eyJ1c2VyIjoidGhlb3J5IiwiZGF0ZSI6IjIwMjQtMDktMTNUMTc6MzI6NTVaIiwidXJpIjoiZGlzdC9wYWlyLzAuMS43L3BhaXItMC4xLjcuemlwIiwiZGlnZXN0cyI6eyJzaGE1MTIiOiJiMzUzYjVhODJiM2I1NGU5NWY0YTI4NTllN2EyYmQwNjQ4YWJjYjM1YTdjMzYxMmIxMjZjMmM3NTQzOGZjMmY4ZThlZTFmMTllNjFmMzBmYTU0ZDdiYjY0YmNmMjE3ZWQxMjY0NzIyYjQ5N2JjYjYxM2Y4MmQ3ODc1MTUxNWI2NyJ9fQ",
+       "signatures": [
+          {
+            "protected":"eyJhbGciOiJSUzI1NiJ9",
+            "header": {"kid": "2024-12-29" },
```
Normally a JWS is like `headerB64.payloadB64.signatureB64`, and therefore I am wondering about why the payload is separate from the header and signature. I believe it is like that in order to support multiple signatures.

Questions to consider are...
- is it necessary to support multiple signatures?
- For 'protected' and 'header', both of these parts typically included in the "header" as described in RFC 7515, so why not combine that into just 'header' as b64 encoded.
- Wondering because if you take the payload, the header (as suggested in previous comment), signature, and concatenate them with "." joined ([RFC7515 section 3.3](https://datatracker.ietf.org/doc/html/rfc7515#section-3.3)), then would that make it exactly like a "normal" signature? i.e. easier to use with existing clients that do signature validation maybe.

theory replied:
It's an official RFC 7515 format called [JWS JSON Serialization](https://www.rfc-editor.org/rfc/rfc7515.html#section-7.2). We don't need URL-safe serialization, so this is just a little friendlier for clients. And yes, it allows multiple signatures, which could be useful for author signing or gradual key rotation.

1. I think it's useful to support multiple signatures because, unlike ephemeral HTTP requests, there's are records at rest and will be around for many years.
2. `protected` and `header` are not combined because the [JWS JSON Serialization](https://www.rfc-editor.org/rfc/rfc7515.html#section-7.2) format separates them.
3. Clients should also support this format, as it's part of the standard. But even if they don't, conversion will be trivial, as you point out.

Really I wanted the payload not to be Base64 URL-encoded, as specified in [JWS-JS](https://datatracker.ietf.org/doc/html/draft-jones-json-web-signature-json-serialization-01), but I believe that spec was abandoned in favor of [RFC 7515 JWS JSON Serialization](https://www.rfc-editor.org/rfc/rfc7515.html#section-7.2), so going with that approved standard, instead.

theory committed:
Reinforce use of JWS JSON Serialization again

sjmiller609 reviewed (approved):
Embedding the public key in clients seems questionable to me, but looks good otherwise.

theory committed:
Reference JWK for key patterns

theory commented:
> Embedding the public key in clients seems questionable to me, but looks good otherwise.

Switched to talking about downloading and using a [JWK Set](https://datatracker.ietf.org/doc/html/rfc7517#section-5) in 01140f3.

theory committed:
Rename release to receipts

theory committed:
Update list of alternative terms

theory committed:
Rename `receits` to `certs`

Still not sure about the name, but I think it's better.

theory committed:
Set RFC number, link to implementation

theory committed:
Rename RFC and cover flattened syntax

Make another edit pass, tweaking some language and structure, but mostly
highlight the support for both the "general" and "flattened" JWS JSON
Serialization syntaxes. Hide lines beginning with `#` in JSON code
fences.

theory committed:
Remove v2

pgguru commented on text/0005-release-certification.md:
````diff
@@ -0,0 +1,497 @@
+distribution `META.json` file:
+
+``` json
+#{
````
What does this mean? Just an extract from a larger JSON object? (The syntax hilighter in GH is complaining about this...)

theory replied:
This `book.toml` config

``` toml
[output.html.code.hidelines]
json = "#"
```

Means that `json` fences will not show lines that begin with `#`. It kinda-sorta works in other parsers, but properly omits them from the rendered view that'll be published on rfcs.pgxn.org.

pgguru commented on text/0005-release-certification.md:
```diff
@@ -0,0 +1,497 @@
+The steps for a client to find, download, and verify a PGXN release would be:
+
+1.  Using a valid PGXN mirror, assemble and fetch the release list for the
+    the extension, `dist/pair.json`.
```
```suggestion
    extension, `dist/pair.json`.
```

theory replied:
Thanks, done in cbc5549.

pgguru commented on text/0005-release-certification.md:
```diff
@@ -0,0 +1,497 @@
+    the extension, `dist/pair.json`.
+2.  Use the release list to determine the best version to install and assemble
+    the URI for its release `META.json`. The format is
+    `dist/{name}/{version}/META.json`; for the above example, that results in
```
```suggestion
    `dist/{name}/{version}/META.json`; for the above example, which results in
```

pgguru replied:
(just sounds better to me)

theory replied:
It's also [more correct](https://www.merriam-webster.com/grammar/when-to-use-that-and-which), thank you!

theory replied:
Fixed in cbc5549.

pgguru commented on text/0005-release-certification.md:
```diff
@@ -0,0 +1,497 @@
+    fails.
+4.  Decode the payload and use its `uri` field to download the release zip
+    file.
+5.  Compare one of the digests from the payload to a digest generated from the
```
This just means there could be multiple types included, like sha-256 or others, we consider a single validation to be sufficient for our purposes here?

theory replied:
A single digest, yes. I suspect there will usually be only one, but wanted to support multiple to simplify migrating SHA-1s. More details in the "Unanswered questions" section.

pgguru commented on text/0005-release-certification.md:
```diff
@@ -0,0 +1,497 @@
+1.  A root key pair is maintained by PGXN, with the private key kept offline.
+2.  A release key pair is generated and signed by the private root key, with
+    the private key kept in an online vault accessible only to PGXN Manager.
+3.  The public keys for both keys are published by PGXN as a [RFC 7517 JWK
```
```suggestion
3.  The public keys for both keypairs are published by PGXN as a [RFC 7517 JWK
```

pgguru replied:
Or `key pairs`, to be consistent with the rest of the document. (This one is like datacenter vs data center for me; the largely accepted spelling doesn't seem correct, so just striving for consistency here.)

theory replied:
Went with `key pairs` in cbc5549.

pgguru commented on text/0005-release-certification.md:
```diff
@@ -0,0 +1,497 @@
+4.  PGXN Manager uses the private release key to sign releases as described
+    above. The most important property in the signed payload is the list of
+    digests.
+5.  Clients **MUST** regularly fetch the [RFC 7517 JWK Set] of public keys,
```
Would a recommended schedule be useful here; weekly? Each invocation of the tool?

theory replied:
Given @sjmiller609's feedback, I'm thinking each invocation, yeah.

pgguru commented on text/0005-release-certification.md:
```diff
@@ -0,0 +1,497 @@
+>     Header Parameter values are integrity protected.
+>
+> *   **header**: The "header" member **MUST** be present and contain the
+>     value JWS Unprotected Header when the JWS Unprotected Header value is
```
```suggestion
>     value "JWS Unprotected Header" when the JWS Unprotected Header value is
```

or maybe

```suggestion
>     value `JWS Unprotected Header` when the JWS Unprotected Header value is
```
also the second non-quoted form here also looks a bit off to me.

theory replied:
These are quoted from [the source](https://datatracker.ietf.org/doc/html/rfc7515#section-7.2.1), which does not quote those strings.

pgguru commented on text/0005-release-certification.md:
```diff
@@ -0,0 +1,497 @@
+>
+> *   **header**: The "header" member **MUST** be present and contain the
+>     value JWS Unprotected Header when the JWS Unprotected Header value is
+>     non- empty; otherwise, it **MUST** be absent. This value is represented
```
```suggestion
>     non-empty; otherwise, it **MUST** be absent. This value is represented
```

theory replied:
Fixed in cbc5549.

pgguru commented on text/0005-release-certification.md:
```diff
@@ -0,0 +1,497 @@
+> conveyed.
+>
+> Additional members can be present in both the JSON objects defined above; if
+> not understood by implementations encountering them, they MUST be ignored.
```
```suggestion
> not understood by implementations encountering them, they **MUST** be ignored.
```

theory replied:
Also quoted from the source.

pgguru commented on text/0005-release-certification.md:
```diff
@@ -0,0 +1,497 @@
+    *   `receipts` Cute, but a bit opaque
+    *   `attestations` Abstract and too JWT-y
+    *   `jws` Dissuades other formats
+    *   `coupons`, `authentication`, `authenticity`, `credentials`,
```
I personally think `certs` -> `certificates`; `authenticity` seems nice to me.

pgguru reviewed (commented):
Mainly nits, a couple q's.

theory committed:
Add fixes thanks to @pgguru
